### PR TITLE
Feature: add Funds and Designations add-on to the form migration process

### PR DESCRIPTION
**Pull request by glaubersilva**
Resolves [GIVE-11] [GIVE-183]

Related PR: https://github.com/impress-org/give-funds/pull/125

## Description

This PR adds the necessary logic to transfer data from the Funds and Designations addon settings to the migrated forms.

This PR also sets the Funds and Designations addon as supported by the visual form builder.

## Affects

The migrated forms.

## Visuals

image

## Testing Instructions

1. Install the Funds and Designations addon from this PR/branch: https://github.com/impress-org/give-funds/pull/125
2. Create a legacy form and set a few fund settings
3. Do the migration process, then check the Funds and Designations block added in the migrated form, you should see in the block settings the settings selected in the previous step.

## Pre-review Checklist

-   [ ] Acceptance criteria satisfied and marked in related issue
-   [ ] Relevant `@unreleased` tags included in DocBlocks
-   [ ] Includes unit tests
-   [ ] Reviewed by the designer (if follows a design)
-   [ ] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

[GIVE-11]: https://stellarwp.atlassian.net/browse/GIVE-11?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[GIVE-183]: https://stellarwp.atlassian.net/browse/GIVE-183?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by jonwaldstein — CHANGES_REQUESTED** [review]

File: src/DonationForms/V2/DonationFormsAdminPage.php [hunk 1]
```diff
@@ -332,7 +332,7 @@ public function getSupportedAddons(): array
             'Salesforce' => defined('GIVE_SALESFORCE_VERSION'),
             'Donation Upsells for WooCommerce' => class_exists('Give_WooCommerce'),
             //            'Manual Donations' => class_exists('Give_Manual_Donations'),
-            //            'Funds' => defined('GIVE_FUNDS_ADDON_NAME'),
+            'Funds' => defined('GIVE_FUNDS_ADDON_NAME'),
```
[the remark this answers is not in the document]
Comment (glaubersilva) [reply src/DonationForms/V2/DonationFormsAdminPage.php:336]: Resolved in 8f2a75296f24539d4abac97830a8bfd4a247f3c9

File: src/FormMigration/Steps/FundsAndDesignations.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+use Give\FormMigration\Contracts\FormMigrationStep;
+use Give\Framework\Blocks\BlockModel;
+
+/**
+ * @unreleased
+ */
+class FundsAndDesignations extends FormMigrationStep
+{
+
```
[the remark this answers is not in the document]
Comment (glaubersilva) [reply src/FormMigration/Steps/FundsAndDesignations.php:23]: Resolved in ba214fafed374ee6f8bdcfc605efcd2d5dca3deb

**Comment by glaubersilva** [comment]
@jonwaldstein Thanks for the suggestions, ready for re-review. =)

File: src/FormMigration/Steps/FundsAndDesignations.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+ * @unreleased
+ */
+class FundsAndDesignations extends FormMigrationStep
+{
+    /**
+     * @unreleased
+     */
+    public function canHandle(): bool
+    {
+        $fundsAndDesignationsAttributes = $this->getFundsAndDesignationsAttributes($this->formV2->id);
```
Comment (jonwaldstein) [inline src/FormMigration/Steps/FundsAndDesignations.php:18]: Another option for the migration step is to add update the `FormDecorator` so you could do something this:

```
return $this->formV2->hasFunds() || $this->formV2->hasFundOptions()
```

Although what you have works, it's a little hard to understand the logic when reading it - so updating would mainly be for comprehension.

File: src/FormMigration/Steps/FundsAndDesignations.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+     * @unreleased
+     */
+    public function process()
+    {
+        $fundsAndDesignationsAttributes = $this->getFundsAndDesignationsAttributes($this->formV2->id);
+
+        if (count($fundsAndDesignationsAttributes['fund']) === 0 &&
+            count($fundsAndDesignationsAttributes['options']) === 0) {
+            return;
+        }
```
Comment (jonwaldstein) [inline src/FormMigration/Steps/FundsAndDesignations.php:34]: since you added the `canHandle` method - this part will be redundant so we can remove.

**Review by jonwaldstein — CHANGES_REQUESTED** [review]

File: src/FormMigration/Steps/FundsAndDesignations.php (hunk 3 above)
Comment (glaubersilva) [reply src/FormMigration/Steps/FundsAndDesignations.php]: Resolve in 78be4079f1976a168446de7d5364219f2f655c7a

File: src/FormMigration/Steps/FundsAndDesignations.php (hunk 4 above)
Comment (glaubersilva) [reply src/FormMigration/Steps/FundsAndDesignations.php]: Resolved in 2f8e345096a9133b59690f70fffeb5e9671bb03b

**Review by jonwaldstein — APPROVED** [review]
Nice job!

**Review by rickalday — APPROVED** [review]
Passed manual QA tests

**Pull request merged**